### DISCUSSION TITLE
How to know when a read is starting and how many bytes were read

### DISCUSSION THREAD

**Discussion by isaacrivriv** [netty/netty#12860] [General]
Hey there! I'm trying to build a custom handler for logging extending the LoggingHandler provided by Netty to add additional logging details for things like reads and writes. Specifically, I'm curious if there's a way to get the amount of bytes that pass through the handler when a read from the socket happens (maybe when readComplete is called?) and also when written to the socket (maybe after calling flush?). Also if there's a way to log when a read is actually starting vs when it is completed in the handler (in channelRead of channelReadComplete). Searching around I haven't found a way to make this happen. Anyone have any advice? Thanks in advance!

**Comment by chrisvest** [comment]
You have to sum up the readable bytes of the ByteBufs that come through channelRead and write, then log and reset in channelReadComplete and flush.
Not sure what you mean by "read is actually starting"; when we call channelRead on the first handler, it's because a `read(2)` or `recv(2)` (or similar) syscall returned data to us.

**Reply by isaacrivriv** [reply]
Thanks for the reply! I think I understand part of what you mentioned. When `channelReadComplete` gets called it is because nothing was read or a half-filled buffer was read so can use the amount read with that call if I understand that correctly. For flush, everything that was written with `channel.write` gets pushed to the socket in a synchronous manner when flush returns? Therefore it would be somewhat the write equivalent of channelReadComplete am I understanding correctly?

For when "read is actually starting" I'm referring to when a call to `channel.read` is called or if autoRead is enabled, then pretty much when a call to the syscall read is about to be performed if there is a way to know when it's about to be called.
